zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -212,3 +215,168 @@ impl<F: FieldExt, const MAX_QUAD_COST_IN_BYTES: usize>
+        let bound_dist = array_init(|_| cb.query_cell());
+        let bound_dist_is_zero = array_init(|idx| {
+            IsZeroGadget::construct(cb, bound_dist[idx].expr())
+        });
+        let lt_gadget =
+            LtGadget::construct(cb, addr_start.expr(), addr_end.expr());
+
+        // Define bound_dist[i] = max(addr_end - addr_start - i, 0)
+        // The purpose of bound_dist is to track if the access to src buffer
+        // is out of bound. When bound_dist[i] == 0, it indicates OOB error
+        // and so bytes[i] has to be 0 correspondingly.
+        // Because the bound_dist is decreasing by at most 1 each time, we can
+        // use this property to reduce the use of LtGadget by adding constraints
+        // to the diff between two consecutive bound_dists.
+
+        // Constraints on bound_dist[0]
+        //   bound_dist[0] == 0 || addr_start + bound_dist[0] == addr_end
+        //   src_addr < src_addr_bound when bound_dist[0] != 0
+        cb.add_constraint(
+            "bound_dist[0] == 0 or addr_start + bound_dist[0] == addr_end",
+            bound_dist[0].expr()
+                * (addr_start.expr() + bound_dist[0].expr() - addr_end.expr()),
+        );
+        cb.add_constraint(
+            "addr_start < addr_end when bound_dist_is_zero[0] == 0",
+            (1.expr() - bound_dist_is_zero[0].expr())
+                * (1.expr() - lt_gadget.expr()),
+        );
+        // Constraints on bound_dist[1..MAX_COPY_BYTES]
+        //   diff = bound_dist[idx - 1] - bound_dist[idx]
+        //   diff == 1 when bound_dist[idx - 1] != 0
+        //   diff == 0 when bound_dist[idx - 1] == 0
+        for idx in 1..MAX_BYTES {
+            let diff = bound_dist[idx - 1].expr() - bound_dist[idx].expr();
+            cb.add_constraint(
+                "diff == 1 when bound_dist[i - 1] != 0",
+                (1.expr() - bound_dist_is_zero[idx - 1].expr())
+                    * (1.expr() - diff.expr()),
+            );
+            cb.add_constraint(
+                "diff == 0 when bound_dist[i - 1] == 0",
+                bound_dist_is_zero[idx - 1].expr() * diff.expr(),
+            )
+        }
+
+        // Constraints on bytes and selectors
+        for i in 0..MAX_BYTES {
+            let selector_prev = if i == 0 {
+                // First selector will always be 1
+                1.expr()
+            } else {
+                selectors[i - 1].expr()
+            };
+            // selector can transit from 1 to 0 only once as [1, 1, 1, ...,
+            // 0, 0, 0]
+            cb.require_boolean(
+                "Constrain selectors can only transit from 1 to 0",
+                selector_prev - selectors[i].expr(),
+            );
+            cb.add_constraint(
+                "bytes[i] == 0 when selectors[i] == 0",
+                (1.expr() - selectors[i].expr()) * bytes[i].expr(),
+            );
+            cb.add_constraint(
+                "bytes[i] == 0 when bound_dist_is_zero[i] == 1",
```

Comment:
do we  also need to check "selectors[i] == 0 when bound_dist_is_zero[i] == 1" ?  
if so, we can omit checking "bytes[i] == 0 when bound_dist_is_zero[i] == 1" since "electors[i] == 0" constraint  can infer "bytes[i] == 0"

Reply:
There could be the case that `selectors[i] == 1` but `bound_dist_is_zero[i] == 1` when the access to buffer is out of bound. In this case, we still need the constrain that `bytes[i] == 0`